epochStart/metachain/rewardsV2.go

```diff
@@ -0,0 +1,496 @@
+	totalTopUpEligible := rc.stakingDataProvider.GetTotalTopUpStakeEligibleNodes()
+	remainingToBeDistributed := rc.economicsDataProvider.RewardsToBeDistributedForBlocks()
+	topUpRewards := rc.computeTopUpRewards(remainingToBeDistributed, totalTopUpEligible)
+	baseRewards := big.NewInt(0).Sub(remainingToBeDistributed, topUpRewards)
+	nbBlocks := big.NewInt(int64(rc.economicsDataProvider.NumberOfBlocks()))
+	baseRewardsPerBlock := big.NewInt(0).Div(baseRewards, nbBlocks)
+	rc.fillBaseRewardsPerBlockPerNode(baseRewardsPerBlock)
+
+	accumulatedDust := big.NewInt(0)
+	baseRewardsPerNode, dust := rc.computeBaseRewardsPerNode(validatorsInfo, baseRewards)
+	accumulatedDust.Add(accumulatedDust, dust)
+	topUpRewardsPerNode, dust := rc.computeTopUpRewardsPerNode(topUpRewards, validatorsInfo)
+	accumulatedDust.Add(accumulatedDust, dust)
+
+	return aggregateBaseAndTopUpRewardsPerNode(baseRewardsPerNode, topUpRewardsPerNode), accumulatedDust
+}
+
+func (rc *rewardsCreatorV2) computeBaseRewardsPerNode(
+	validatorsInfo map[uint32][]*state.ValidatorInfo,
+	baseRewards *big.Int,
+) (map[uint32][]*big.Int, *big.Int) {
+	baseRewardsPerNode := make(map[uint32][]*big.Int)
+	accumulatedRewards := big.NewInt(0)
+
+	for shardID, valInfoList := range validatorsInfo {
+		baseRewardsPerNode[shardID] = make([]*big.Int, len(valInfoList))
+		for i, valInfo := range valInfoList {
+			baseRewardsPerNode[shardID][i] = big.NewInt(0).Mul(
+				rc.mapBaseRewardsPerBlockPerValidator[shardID],
+				big.NewInt(0).SetInt64(int64(valInfo.NumSelectedInSuccessBlocks)))
+			accumulatedRewards.Add(accumulatedRewards, baseRewardsPerNode[shardID][i])
+		}
+	}
+
+	return baseRewardsPerNode, big.NewInt(0).Sub(baseRewards, accumulatedRewards)
+}
+
+func (rc *rewardsCreatorV2) computeTopUpRewardsPerNode(
+	topUpRewards *big.Int,
+	nodesInfo map[uint32][]*state.ValidatorInfo,
+) (map[uint32][]*big.Int, *big.Int) {
+
+	topUpRewardsPerNode := make(map[uint32][]*big.Int)
+	accumulatedTopUpRewards := big.NewInt(0)
+	topUpPerNode := rc.getTopUpForAllEligibleNodes(nodesInfo)
+	topUpRewardPerShard := rc.computeTopUpRewardsPerShard(topUpRewards, topUpPerNode)
+	totalPowerInShard, nodesPower := computeNodesPowerInShard(nodesInfo, topUpPerNode)
+
+	// topUpRewardPerNodeInShardX = nodePowerInShardX*topUpRewardsShardX/totalPowerInShardX
+	for shardID, nodesPowerList := range nodesPower {
+		topUpRewardsPerNode[shardID] = make([]*big.Int, len(nodesPowerList))
+		if totalPowerInShard[shardID].Cmp(big.NewInt(0)) == 0 {
+			log.Warn("rewardsCreatorV2.computeTopUpRewardsPerNode",
+				"error", "shardPower zero",
+				"shardID", shardID)
+		}
+
+		for i, nodePower := range nodesPowerList {
+			if totalPowerInShard[shardID].Cmp(big.NewInt(0)) == 0 {
+				// avoid division by zero
+				topUpRewardsPerNode[shardID][i] = big.NewInt(0)
+				continue
+			}
+
+			topUpRewardsPerNode[shardID][i] = big.NewInt(0).Mul(nodePower, topUpRewardPerShard[shardID])
```

Comment:
I think it would be better to have a structure like map[blsKey]struct{topUp, baseReward) instead of maps and lists. It will be hard to debug.

Reply:
You can add directly to state.ValidatorInfo, 2 new fields.

Reply:
will do this in a separate PR, together with the other unit tests tests

Reply:
done